packages/platform-core/API.md

```diff
@@ -7,6 +7,64 @@
 import { BackendIdentifier } from '@aws-amplify/plugin-types';
 import z from 'zod';
 
+// @public
+export class AmplifyError extends Error {
+    constructor(name: AmplifyErrorType, classification: AmplifyErrorClassification, options: AmplifyErrorOptions, downstreamError?: Error | undefined);
+    // (undocumented)
+    readonly classification: AmplifyErrorClassification;
+    // (undocumented)
+    readonly code?: string;
+    // (undocumented)
+    readonly details?: string;
+    // (undocumented)
+    readonly downstreamError?: Error | undefined;
+    // (undocumented)
+    static fromError: (error: unknown) => AmplifyError;
+    // (undocumented)
+    static fromStderr: (_stderr: string) => AmplifyError | undefined;
+    // (undocumented)
+    readonly link?: string;
+    // (undocumented)
+    readonly message: string;
+    // (undocumented)
+    readonly name: AmplifyErrorType;
+    // (undocumented)
+    readonly resolution?: string;
+    // (undocumented)
+    serializedError?: string;
+}
+
+// @public
+export type AmplifyErrorClassification = 'FAULT' | 'ERROR';
+
+// @public
+export type AmplifyErrorOptions = {
+    message: string;
+    details?: string;
+    resolution?: string;
+    link?: string;
+    code?: string;
+};
+
+// @public
+export type AmplifyErrorType = AmplifyUserErrorType | AmplifyLibraryFaultType;
+
+// @public
+export class AmplifyFault extends AmplifyError {
+    constructor(name: AmplifyLibraryFaultType, options: AmplifyErrorOptions, downstreamError?: Error);
+}
+
+// @public
+export type AmplifyLibraryFaultType = 'UnknownFault';
+
+// @public
+export class AmplifyUserError extends AmplifyError {
+    constructor(name: AmplifyUserErrorType, options: AmplifyErrorOptions, downstreamError?: Error);
+}
+
+// @public
+export type AmplifyUserErrorType = 'InvalidSchemaAuthError' | 'InvalidSchemaError' | 'ExpiredTokenError' | 'CloudFormationDeploymentError' | 'CFNUpdateNotSupportedError' | 'SyntaxError' | 'BackendBuildError' | 'BootstrapNotDetectedError' | 'AccessDeniedError' | 'FileConventionError';

```

Comment:
I think we should decentralize this and have each package provide error types.

For example 'CloudFormationDeploymentError' .
If I'm a component that is handling this error I should already depend on cdk deployer.
On the other hand if I'm not interested in type (other than opaque printing/passing to telemetry) the enumeration and actual value should matter for me.

Reply:
by that I mean a component that's trying to :
```
catch (err: AmplifyUserError) {
 if (err.name === 'CloudFormationDeploymentError') {
  // attempt some recovery 
 }
}
```

while writing above snippet - should parameter be called `type` not `name` if enumeration type has `Type` suffix ?

Also. Should `name` be that strongly typed or should we just allow `string` there and let components set whatever they want.
Other option to decentralize this is to make `name: T` (template type) and let components create factory or subtypes of errors.

The problem with centralized definition of "error type" surfaced in gen 1 where coordinating change between cli and api repos was taking unnecessary long time and releases. We shouldn't assume that all components that would construct `AmplifyErrors` are always going to be in single mono repo.

Reply:
(solving this problem might be punted for later as release blocker, because changes in this interface are going to be quite disruptive if deferred further).

Reply:
yes perhaps this is a `code` or `errorType`. The challenge remains that the verticals need to throw errors that the platform can easily categorize as 4xx/5xx and perhaps further. In Gen1 that was tied to the name having the word `Error` in it. In Gen2, I envision that we publish `AmplifyError.classification` as well to the telemetry.

Whether we need `name` or not, I believe we don't. The required message should capture the free form input from throwers.

If you are ok, I'll create a release blocker backlog as this needs a slightly more thoughts put into it.

Reply:
That means we need two properties on error where one is granular and other is classification.
I'd assign catogorization at throw call site, and not try to classify this in platform. If initial classification is wrong then component that's catching error can wrap it and reclassify.

Feel free to cut release blocker.

Reply:
That is currently the case if the throw site uses `AmplifyError` instance since `name` is mandatory. The biggest hurdle is when call site throws generic `Error`.

Created issue here https://github.com/aws-amplify/amplify-backend/issues/748